addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,80 +193,69 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
-        name = product.display_name
-        if product.description_purchase:
-            name += '\n' + product.description_purchase
+        price_unit, taxes = self._purchase_service_get_price_unit_and_taxes(supplierinfo, purchase_order)
+        name = self._purchase_service_get_product_name(supplierinfo, purchase_order, quantity)
 
         return {
             'name': name,
             'product_qty': purchase_qty_uom,
             'product_id': self.product_id.id,
             'product_uom': self.product_id.uom_po_id.id,
             'price_unit': price_unit,
-            'date_planned': fields.Date.from_string(purchase_order.date_order) + relativedelta(days=int(supplierinfo.delay)),
+            'date_planned': purchase_order.date_order + relativedelta(days=int(supplierinfo.delay)),
             'taxes_id': [(6, 0, taxes.ids)],
             'order_id': purchase_order.id,
             'sale_line_id': self.id,
         }
 
+    def _purchase_service_match_supplier(self, warning=True):
+        # determine vendor of the order (take the first matching company and product)
+        suppliers = self.product_id._select_seller(quantity=self.product_uom_qty, uom_id=self.product_uom)
+        if warning and not suppliers:
+            raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (self.product_id.display_name,))
+        return suppliers[0]
+
+    def _purchase_service_match_purchase_order(self, partner, company=False):
+        return self.env['purchase.order'].search([
+            ('partner_id', '=', partner.id),
+            ('state', '=', 'draft'),
+            ('company_id', '=', (company if company else self.company_id).id),
+        ], limit=1)
+
+    def _create_purchase_order(self, supplierinfo):
+        values = self._purchase_service_prepare_order_values(supplierinfo)
+        return self.env['purchase.order'].with_context(mail_create_nosubscribe=True).create(values)
+
+    def _match_or_create_purchase_order(self, supplierinfo):
+        purchase_order = self._purchase_service_match_purchase_order(supplierinfo.partner_id)
+        if not purchase_order:
+            purchase_order = self._create_purchase_order(supplierinfo)
+        return purchase_order
+
     def _purchase_service_create(self, quantity=False):
         """ On Sales Order confirmation, some lines (services ones) can create a purchase order line and maybe a purchase order.
             If a line should create a RFQ, it will check for existing PO. If no one is find, the SO line will create one, then adds
             a new PO line. The created purchase order line will be linked to the SO line.
             :param quantity: the quantity to force on the PO line, expressed in SO line UoM
         """
-        PurchaseOrder = self.env['purchase.order']
         supplier_po_map = {}
         sale_line_purchase_map = {}
+        purchase_lines = self.env['purchase.order.line']
```

Comment:
In first commit, `return purchase_lines` is the only usage of this variable and it is dead code (after a `return` statement).
```suggestion
```